docs/rfc/0001_ctrlkit.md

```diff
@@ -0,0 +1,325 @@
+* Too many polyfill codes to write. When we want to request subresources of another resource, we have to write a method that invokes the client of apiserver with some selectors, e.g., names and labels. Writing these polyfill codes is painful if there are many kinds of subresources.
+# **Related Resources**
+
+Some concepts and projects:
+
+* [Kubernetes -- Custom resources](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/)
+* [Kubernetes -- Operator pattern](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/)
+* [kubernetes-sigs/controller-runtime](https://github.com/kubernetes-sigs/controller-runtime)
+* [pingcap/tidb-operator](https://github.com/pingcap/tidb-operator)
+* [cockroachdb/cockroach-operator](https://github.com/cockroachdb/cockroach-operator)
+* [ApsaraDB/polardbx-operator](https://github.com/ApsaraDB/galaxykube)
+
+
+# **Explanation**
+
+## **Detailed design**
+
+### **Overview**
+
+The framework consists of 3 parts:
+* A DSL to define the state and actions
+* A command-line tool to generate the polyfill codes
+* A library to help organize the workflow and provide utilities
+
+### **Concepts**
+
+**State**: Kubernetes resources obtained from apiserver represent the reality that the operator observes. Typically we query our desired resources by getting by a namespaced name or by listing with label/field selectors.
+
+**Action**: The basic unit of workflow. The [Single Responsibility Principle](https://en.wikipedia.org/wiki/Single-responsibility_principle) should also be followed while designing the actions. Actions might take some states as input. The output of the actions affects the behavior of the event queue.
+
+**Workflow**: A collection of actions that are organized in a graph.
+
+### **DSL**
+
+The DSL is for declaring the manager of some CR that contains states and actions. DSL also contains statements to help the tool generate the polyfill codes.
+
+1. `bind GV Package`
+    * `bind` a group/version (e.g., apps/v1) to a Go package (e.g., `k8s.io/api/apps/v1`), which provides the types in GV.
+2. `alias Kind GVK`
+    * `alias` a group/version/kind to a single word name, commonly the same name with type, to make the reference of GVK easier.
+3. `decl ControllerManager for Kind` block
+    * `decl` a controller manager for a specific kind, with a body as follows:
+    * `state` block, with a body:
+      * `name Kind` / `name []Kind` block defines a state's name and kind (where `[]Kind` means a list of resources), with a body:
+        * Each line is a selector. Supported selectors and the effects are summarized in the table below.
+    * `action` block, with a body:
+      * `ActionName(states...)` defines an action with its name and states it requires. The `states...` part can be empty. But if there's any state required, it must be declared before in some `state` block.
```

Comment:
Does this mean that an action would be done if the target resource is any of the `states` list?

Reply:
No. The `states` here are mainly k8s resources that affect how the action works besides the target. Some actions don't rely on k8s resources. For example, they might only react to meta-service changes, such as `SyncCatalogFromMetaIntoStatus()`.